### PR TITLE
[nexus] Support allocating external IP addresses for internal services

### PR DESCRIPTION
**Pull request by smklein**
Extends the allocation of external IPs to also support "service IPs", which are not associated with instances.

- Modifies the `instance_external_ip` table to let the project be optional. A "service IP" is one where the project and instance IDs are both "NULL".
- Modifies the external IP allocation CTE to support this new type of provision.
  - While I was there, fixed the idempotence of provisioniong when out of IP pool space for all external IP allocations.

Part of https://github.com/oxidecomputer/omicron/issues/1530
Fixes https://github.com/oxidecomputer/omicron/issues/1610

### CODE REVIEW

File: nexus/src/db/queries/external_ip.rs [hunk 1]
```diff
@@ -134,8 +135,8 @@ const MAX_PORT: i32 = u16::MAX as _;
 ///     -- possible on replay of a saga node.
 ///     INSERT INTO
 ///         instance_external_ip
-///     (SELECT * FROM external_ip)
-///     ON CONFLICT
+///     (SELECT * FROM next_external_ip)
+///     ON CONFLICT (id)
```
Comment (smklein) [inline nexus/src/db/queries/external_ip.rs:139]: I didn't actually change the code here, just updated the documentation of this comment.

File: nexus/src/db/queries/external_ip.rs [hunk 2]
```diff
@@ -437,18 +442,21 @@ impl NextExternalIp {
...
-            out.push_bind_param::<sql_types::Uuid, Uuid>(pool_id)?;
-        } else {
-            out.push_sql("(");
-            out.push_identifier(dsl::project_id::NAME)?;
-            out.push_sql(" = ");
-            out.push_bind_param::<sql_types::Uuid, Uuid>(self.ip.project_id())?;
-            out.push_sql(" OR ");
-            out.push_identifier(dsl::project_id::NAME)?;
-            out.push_sql(" IS NULL)");
+        match self.ip.source() {
```
Comment (smklein) [inline nexus/src/db/queries/external_ip.rs:445]: This is the callsite which motivated the creation of `IpSource`. Without that structure, we'd have an `Option<pool_id>` and an `Option<project_id>`. Now that the `project_id` is optional, *both* could be null. In that case, this query would be invalid.

File: nexus/src/db/datastore/instance_external_ip.rs [hunk 3]
```diff
@@ -20,6 +20,7 @@ use crate::db::update_and_check::UpdateStatus;
 use async_bb8_diesel::AsyncRunQueryDsl;
 use chrono::Utc;
 use diesel::prelude::*;
+use nexus_types::identity::Resource;
```
Comment (smklein) [inline nexus/src/db/datastore/instance_external_ip.rs:23]: There's a comment on line 33, referring to `allocate_instance_snat_ip`, saying:

> // TODO-correctness: This should be made idempotent.

I think it is? Am I clear to remove this TODO?

File: common/src/sql/dbinit.sql [hunk 4]
```diff
@@ -1105,7 +1105,13 @@ CREATE TYPE omicron.public.ip_kind AS ENUM (
...
+    'floating',
+
+    /*
+     * A service IP is an IP address not attached to a project nor an instance.
+     * It's intended to be used for internal services.
+     */
+    'service'
 );
 
 /*
```
Comment (bnaecker) [inline common/src/sql/dbinit.sql:1117]: This comment is actually intended to be attached to the `CREATE TABLE` statement, GitHub just apparently doesn't allow comments on lines that are not on or near changes in the actual PR.

At some point, doesn't have to be now, we probably want to rename this table to just `external_ip`. It's no longer just associated with addresses for guests.

Comment (smklein) [reply common/src/sql/dbinit.sql:1117]: I'll do that in a follow-up PR to this change -- hopefully that keeps the "simple change" piece of renaming separate from this PR, which does have some more subtle changes.

https://github.com/oxidecomputer/omicron/issues/1614

File: common/src/sql/dbinit.sql [hunk 5]
```diff
@@ -1132,7 +1138,7 @@ CREATE TABLE omicron.public.instance_external_ip (
     ip_pool_range_id UUID NOT NULL,
 
     /* FK to the `project` table. */
```
Comment (bnaecker) [inline common/src/sql/dbinit.sql:1140]: Since this is now nullable, I think it'd be worthwhile to add constraints on it like we have for `instance_id`. It should be pretty similar, but not identical, only allowing nulls here if this is a service IP address.

File: nexus/src/db/datastore/instance_external_ip.rs (hunk 3 above)
Comment (bnaecker) [reply nexus/src/db/datastore/instance_external_ip.rs:23]: I think that's right, yep

File: nexus/src/db/queries/external_ip.rs (hunk 2 above)
Comment (bnaecker) [reply nexus/src/db/queries/external_ip.rs:445]: I think I understand your point. For instance-related IP addresses, we always have some project, and possibly a pool. For service-related IP addresses we never have a project, and always have a pool. (I think this just restates the invariants maintained by the constructors like `InstanceExternalIp::for_floating` etc.)

I feels a bit weird to me to have two fields though, the project and this new source field, because they're sometimes duplicates. What if we combine the pool and project (and maybe the kind) into a single field that captures these? Something like this, though the name is not great now:

```rust
pub enum IpSource {
    Instance { project_id: Uuid, pool_id: Option<Uuid> },
    Service { pool_id: Uuid },
}
```

I say "maybe the kind" because we now have some duplication there too. It might be better to just have this enum with a variant for each kind, where the instance-related ones all have the same fields. I dunno, WDYT?

File: nexus/src/db/queries/external_ip.rs [hunk 6]
```diff
@@ -966,7 +1002,161 @@ mod tests {
...
+        let err = context
+            .db_datastore
+            .allocate_service_ip(&context.opctx, id3, rack_id)
+            .await
+            .expect_err("Should have failed to allocate after pool exhausted");
+        assert_eq!(
+            err,
+            Error::InvalidRequest {
+                message: String::from(
+                    // TODO: The error is a bit misleading; this isn't an IP
```
Comment (bnaecker) [inline nexus/src/db/queries/external_ip.rs:1064]: Fair, it could just say "No External IP addresses available".

**Review by bnaecker — COMMENTED** [review]
Thanks for finding / fixing the idempotency edge case, and the tests look great! I've got a few questions around constraints and the new `IpSource` field.

File: nexus/src/db/queries/external_ip.rs (hunk 6 above)
Comment (bnaecker) [reply nexus/src/db/queries/external_ip.rs:1064]: Actually, thinking out loud, will users ever see this error message in the case that we fail to get an address for an Oxide service? For the most part, they won't be deciding to create new services like Nexus, the control plane itself should do that. One potential exception is during setup and bootstrapping, where we may need to alert the operator that we couldn't bring up enough Nexus instances (or any service) because there weren't enough IPs. Not sure.

File: nexus/src/db/queries/external_ip.rs (hunk 2 above)
Comment (smklein) [reply nexus/src/db/queries/external_ip.rs:453]: I like this proposal. I had to shuffle around some of the query construction, but the structures make more sense when specified within the enum that you described.

Updated!

File: nexus/src/db/datastore/instance_external_ip.rs (hunk 3 above)
Comment (smklein) [reply nexus/src/db/datastore/instance_external_ip.rs:23]: Removed

File: common/src/sql/dbinit.sql (hunk 5 above)
Comment (smklein) [reply common/src/sql/dbinit.sql:1156]: Done - I added a constraint to the `project_id` specifically

File: nexus/src/db/queries/external_ip.rs (hunk 6 above)
Comment (smklein) [reply nexus/src/db/queries/external_ip.rs:1078]: I think just omitting the "for new instance" leaves the mechanics accurate. If a user sees an unwrapped error that looks like:

```
- Failed to allocate instance
- No external IP addresses available
```

That's perfectly understandable, IMO. I think it's risky to make error messages dependent on the caller's context, which we don't know ahead of time.

Comment (smklein) [reply nexus/src/db/queries/external_ip.rs:1078]: BTW, this error message is updated in https://github.com/oxidecomputer/omicron/pull/1635

**Comment by smklein** [comment]
> Thanks for finding / fixing the idempotency edge case, and the tests look great! I've got a few questions around constraints and the new `IpSource` field.

All comments addressed - this should be ready for re-review!

File: nexus/db-model/src/external_ip.rs [hunk 7]
```diff
@@ -78,6 +79,18 @@ impl From<InstanceExternalIp> for sled_agent_client::types::SourceNatConfig {
...
 }
 
+/// Describes where the IP candidates for allocation come from: either
+/// from an IP pool, or from a project.
+///
+/// This ensures that a source is always specified, and a caller cannot
+/// request an external IP allocation without providing at least one of
+/// these options.
+#[derive(Debug, Clone, Copy)]
+pub enum IpSource {
```
Comment (bnaecker) [inline nexus/db-model/src/external_ip.rs:89]: Nice, this makes it much easier to see the situations under which each of the pool/project IDs are required or not. Thanks!

**Review by bnaecker — APPROVED** [review]
Excellent, thanks for taking another pass. Looks great!
